MatrixSDK/Background/MXBackgroundSyncService.swift

```diff
@@ -0,0 +1,521 @@
+                NSLog("[MXBackgroundSyncService] fetchEvent: Event already decrypted.")
+                Queues.dispatchQueue.async {
+                    completion(.success(event))
+                }
+                return
+            }
+            
+            //  should decrypt it first
+            if canDecryptEvent(event) {
+                //  we have keys to decrypt the event
+                NSLog("[MXBackgroundSyncService] fetchEvent: Event needs to be decrpyted, and we have the keys to decrypt it.")
+                
+                do {
+                    try decryptEvent(event)
+                    Queues.dispatchQueue.async {
+                        completion(.success(event))
+                    }
+                } catch let error {
+                    NSLog("[MXBackgroundSyncService] fetchEvent: Decryption failed even crypto claimed it has the keys.")
+                    handleDecryptionFailure(withError: error)
+                }
+            } else {
+                //  we don't have keys to decrypt the event
+                NSLog("[MXBackgroundSyncService] fetchEvent: Event needs to be decrpyted, but we don't have the keys to decrypt it.")
+                handleDecryptionFailure(withError: nil)
+            }
+        }
+        
+        //  check if we've fetched the event before
+        if let cachedEvent = self.cachedEvents[eventId] {
+            //  use cached event
+            handleEncryption(forEvent: cachedEvent)
+        } else {
+            //  do not call the /event api and just check if the event exists in the store
+            let event = syncResponseStore.event(withEventId: eventId, inRoom: roomId) ?? store.event(withEventId: eventId, inRoom: roomId)
+            
+            if let event = event {
+                NSLog("[MXBackgroundSyncService] fetchEvent: We have the event in stores.")
+                //  cache this event
+                self.cachedEvents[eventId] = event
+                
+                //  handle encryption for this event
+                handleEncryption(forEvent: event)
+            } else if allowSync {
+                NSLog("[MXBackgroundSyncService] fetchEvent: We don't have the event in stores. Launch a background sync to fetch it.")
+                self.launchBackgroundSync(forEventId: eventId, roomId: roomId, completion: completion)
+            } else {
+                NSLog("[MXBackgroundSyncService] fetchEvent: We don't have the event in stores. Do not sync anymore.")
+                Queues.dispatchQueue.async {
+                    completion(.failure(MXBackgroundSyncServiceError.unknown))
+                }
+            }
+        }
+    }
+    
+    private func launchBackgroundSync(forEventId eventId: String,
+                                      roomId: String,
+                                      completion: @escaping (MXResponse<MXEvent>) -> Void) {
+        guard let eventStreamToken = store.eventStreamToken else {
```

Comment:
Can the service use its own stream token? ie, syncResponseStore.syncResponse.nextBatch?

The goal is to separate MXSession store and MXBgSyncService as much as possible. Writing to the MXSession store from here adds weakness to MXSession.
If there is a crash while storing MXBgSyncService data but MXStore.eventStreamToken gets stored from here, MXSession will have a hole in its data.

Reply:
Updated to read first from syncResponse, but we still should continue to read from store if syncResponse is not present.